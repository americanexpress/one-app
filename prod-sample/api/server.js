/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

const fs = require('fs');
const https = require('https');
const http = require('http');
const jsonServer = require('json-server');
const pause = require('connect-pause');

const server = jsonServer.create();
const router = jsonServer.router('db.json');
const middlewares = jsonServer.defaults();

const delay = process.argv[2] ? Number(process.argv[2]) : 0;
const options = {
  key: fs.readFileSync('./api-privkey.pem'),
  cert: fs.readFileSync('./api-cert.pem'),
};

if (delay !== 0) {
  server.use(pause(delay));
}

server.use(middlewares);
server.use((req, res, next) => {
  const secretMessage = req.headers['auth-token'] ? 'you are being watched' : 'unauthorised';
  res.append('secret-message', secretMessage);
  res.append('traceparent', req.headers.traceparent);
  next();
});
server.use(router);

https.createServer(options, server).listen(443, '0.0.0.0');
http.createServer({}, server).listen(80, '0.0.0.0');
