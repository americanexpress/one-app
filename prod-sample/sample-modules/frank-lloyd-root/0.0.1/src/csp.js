/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import contentSecurityPolicyBuilder from 'content-security-policy-builder';
import ip from 'ip';

export default contentSecurityPolicyBuilder({
  directives: {
    reportUri: process.env.ONE_CLIENT_CSP_REPORTING_URL,
    defaultSrc: [
      "'self'",
    ],
    scriptSrc: [
      "'unsafe-inline'",
      "'self'",
      // used for our sample app deployment in heroku
      'https://one-app-statics.surge.sh',
      // used by integration tests running in docker where domain names are aliased
      'https://sample-cdn.frank',
      // used for local development
      `${ip.address()}:${process.env.HTTP_ONE_APP_DEV_CDN_PORT || 3001}`,
      // used for local development
      `localhost:${process.env.HTTP_ONE_APP_DEV_CDN_PORT || 3001}`,
    ],
    imgSrc: [
      "'self'",
      // used for our sample app deployment in heroku
      'https://one-app-statics.surge.sh',
      // used by integration tests running in docker where domain names are aliased
      'https://sample-cdn.frank',
    ],
    styleSrc: [
      "'self'",
      "'unsafe-inline'",
    ],
    connectSrc: [
      "'self'",
      '*.api.frank',
      // used for our sample app deployment in heroku
      'https://one-app-statics.surge.sh',
      // used for local development
      `${ip.address()}:${process.env.HTTP_ONE_APP_DEV_CDN_PORT || 3001}`,
      // used for local development
      `localhost:${process.env.HTTP_ONE_APP_DEV_CDN_PORT || 3001}`,
      // used by integration tests running in docker where domain names are aliased
      'https://sample-cdn.frank',
    ],
  },
});
