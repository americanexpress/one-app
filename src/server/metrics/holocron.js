/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createCounter } from './counters';
import { createGauge } from './gauges';

import createMetricNamespace from './create-metric-namespace';

const holocronNamespace = createMetricNamespace('holocron');

createCounter({
  name: holocronNamespace('module_map_poll', 'total'),
  help: 'total times the module map has been polled',
});

createCounter({
  name: holocronNamespace('module_map_updated', 'total'),
  help: 'total times the module map has been updated',
});

createCounter({
  name: holocronNamespace('module_map_poll_restarted', 'total'),
  help: 'total times the module map poll has been restarted',
});

createGauge({
  name: holocronNamespace('module_map_poll_wait', 'seconds'),
  help: 'delay until the next poll of the module map',
});

createGauge({
  name: holocronNamespace('module_map_poll_consecutive_errors', 'total'),
  help: 'how many times has the server encountered errors polling the module map',
});

export default holocronNamespace.getMetricNames();
