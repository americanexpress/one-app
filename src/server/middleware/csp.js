/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { v4 as uuidV4 } from 'uuid';
import ip from 'ip';

function parsePolicy(headerValue) {
  const policy = {};
  headerValue
    .split(';')
    .map((s) => s.trim())
    .forEach((directive) => {
      if (directive) {
        const parts = /^([^ ]+) (.+)$/.exec(directive);
        if (parts) {
          const directiveName = parts[1];
          policy[directiveName] = parts[2].split(' ');
        } else {
          policy[directive] = true;
        }
      }
    });
  return policy;
}

const insertSource = (policy, directive, value) => {
  const policyContainsDirective = policy.search(directive) !== -1;
  const updatedPolicy = policyContainsDirective
    ? policy.replace(directive, `${directive} ${value}`)
    : policy;
  return updatedPolicy;
};

export const cspCache = {
  policy: "default-src 'none';",
};

export function updateCSP(csp) {
  cspCache.policy = csp;
}

export function getCSP() {
  return parsePolicy(cspCache.policy);
}

const csp = () => (req, res, next) => {
  const { policy } = cspCache;
  const scriptNonce = uuidV4();
  let updatedPolicy;
  if (process.env.NODE_ENV === 'development') {
    const updatedScriptSrc = insertSource(policy, 'script-src', `'nonce-${scriptNonce}' ${ip.address()}:${process.env.HTTP_ONE_APP_DEV_CDN_PORT} localhost:${process.env.HTTP_ONE_APP_DEV_CDN_PORT}`);
    updatedPolicy = insertSource(
      updatedScriptSrc,
      'connect-src',
      [
        `${ip.address()}:${process.env.HTTP_ONE_APP_DEV_CDN_PORT}`,
        `localhost:${process.env.HTTP_ONE_APP_DEV_CDN_PORT}`,
        `${ip.address()}:${process.env.HTTP_ONE_APP_DEV_PROXY_SERVER_PORT}`,
        `localhost:${process.env.HTTP_ONE_APP_DEV_PROXY_SERVER_PORT}`,
      ].join(' ')
    );
  } else {
    updatedPolicy = insertSource(policy, 'script-src', `'nonce-${scriptNonce}'`);
  }

  res.scriptNonce = scriptNonce;
  res.setHeader('Content-Security-Policy', updatedPolicy);
  next();
};

export default csp;
