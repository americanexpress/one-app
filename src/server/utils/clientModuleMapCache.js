/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

let cache = {};

/* Filters bundle types so only the required client moduleBundleType is returned
in ../server/middleware/sendHtml.js and reducing the html payload size */
function filterBundles(moduleMap, moduleBundleType) {
  return {
    ...moduleMap,
    modules: Object.entries(moduleMap.modules).reduce((acc, [moduleName, moduleBundles]) => (
      {
        ...acc,
        [moduleName]: {
          baseUrl: moduleBundles[moduleBundleType].url.replace(/[^/]+\.js$/i, ''),
          [moduleBundleType]: moduleBundles[moduleBundleType],
        },
      }
    ), {}),
  };
}

export function setClientModuleMapCache(moduleMap) {
  cache = {
    browser: filterBundles(moduleMap, 'browser'),
    legacyBrowser: filterBundles(moduleMap, 'legacyBrowser'),
  };
}

export function getClientModuleMapCache() {
  return cache;
}
