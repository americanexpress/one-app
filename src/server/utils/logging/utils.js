/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { STATUS_CODES } from 'http';

import chalk from 'chalk';

const coloredLevels = {
  error: chalk.red('error'),
  warn: chalk.yellow('warn'),
  log: chalk.blue('log'),
  info: chalk.gray('info'),
};

function printStatusCode(obj) {
  const { request: { statusCode } } = obj;

  if (statusCode === null) {
    return chalk.black.bgRed(statusCode);
  }

  if (statusCode < 300) {
    // 100s, 200s
    return chalk.green(statusCode);
  }
  if (statusCode < 400) {
    // 300s
    return chalk.blue(statusCode);
  }
  if (statusCode < 500) {
    // 400s
    return chalk.yellow(statusCode);
  }
  // 500s
  return chalk.red(statusCode);
}

function printStatusMessage(obj) {
  const { request: { statusText, statusCode } } = obj;

  return statusText || STATUS_CODES[statusCode] || chalk.black.bgRed('Timed Out?');
}

function printDurationTime(obj) {
  const { request: { timings: { duration } } } = obj;
  if (duration < 100) {
    return chalk.green(duration);
  }
  if (duration < 1e3) {
    return chalk.yellow(duration);
  }
  if (duration < 6e3) {
    return chalk.red(duration);
  }
  return chalk.black.bgRed(duration);
}

export {
  coloredLevels,
  printStatusCode,
  printStatusMessage,
  printDurationTime,
};
