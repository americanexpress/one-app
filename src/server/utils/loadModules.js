/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { getModule } from 'holocron';
import { updateModuleRegistry } from 'holocron/server';

import onModuleLoad, { CONFIGURATION_KEY } from './onModuleLoad';
import batchModulesToUpdate from './batchModulesToUpdate';
import getModulesToUpdate from './getModulesToUpdate';
import { getServerStateConfig } from './stateConfig';
import { setClientModuleMapCache } from './clientModuleMapCache';
import { updateCSP } from '../middleware/csp';

const loadModules = async () => {
  const moduleMapResponse = await fetch(process.env.HOLOCRON_MODULE_MAP_URL);
  const moduleMap = await moduleMapResponse.json();
  const serverConfig = getServerStateConfig();

  const loadedModules = await updateModuleRegistry({
    moduleMap,
    batchModulesToUpdate,
    onModuleLoad,
    getModulesToUpdate,
  });

  const loadedModuleNames = Object.keys(loadedModules);

  if (loadedModuleNames.length > 0) {
    setClientModuleMapCache(moduleMap);
  }

  const rootModuleLoaded = loadedModuleNames.includes(serverConfig.rootModuleName);

  if (!rootModuleLoaded) {
    return loadedModules;
  }

  const RootModule = getModule(serverConfig.rootModuleName);
  const { [CONFIGURATION_KEY]: { csp } = {} } = RootModule;
  updateCSP(csp);

  return loadedModules;
};

export default loadModules;
