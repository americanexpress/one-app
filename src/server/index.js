/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import path from 'path';
import fs from 'fs';
import Intl from 'lean-intl';
import enData from 'lean-intl/locale-data/json/en.json';

import './init';
// Allow env config to run before importing holocron. (x2)
import ssrServer from './ssrServer';
import metricsServer from './metricsServer';
import listen from './listen';
import { addServer, shutdown } from './shutdown';
import pollModuleMap from './utils/pollModuleMap';
import loadModules from './utils/loadModules';

function ssrServerStart() {
  // need to load _some_ locale so that react-intl does not prevent modules from loading
  // eslint-disable-next-line no-underscore-dangle
  Intl.__addLocaleData(enData);
  return loadModules()
    .then(() => new Promise((res, rej) => {
      addServer(listen(ssrServer, (err, { port }) => {
        if (err) {
          rej(err);
        } else {
          console.log(`ðŸŒŽ One App server listening on port ${port}`);
          res();
        }
      }));
    }))
    .then(pollModuleMap);
}

function metricsServerStart() {
  return new Promise((res, rej) => {
    const port = process.env.HTTP_METRICS_PORT;
    addServer(metricsServer.listen(port, (err) => (err ? rej(err) : res(port))));
  })
    .then(
      (port) => console.log(`ðŸ“Š Metrics server listening on port ${port}`),
      (err) => {
        console.error('error encountered starting the metrics server', err);
        throw err;
      }
    );
}

function appServersStart() {
  return Promise.all([
    ssrServerStart(),
    metricsServerStart(),
  ]);
}

let serverChain;
if (process.env.NODE_ENV === 'development') {
  const pathToDevEndpoints = path.join(process.cwd(), '.dev', 'endpoints', 'index.js');

  const getRemotesFromDevEndpointsFile = () => {
    const moduleRemotes = {};
    if (fs.existsSync(pathToDevEndpoints)) {
      // eslint-disable-next-line global-require,import/no-dynamic-require
      Object.values(require(pathToDevEndpoints)()).forEach(({ destination, devProxyPath }) => {
        moduleRemotes[devProxyPath] = destination;
      });
      return moduleRemotes;
    }
    return {};
  };

  /* eslint
    global-require: 0,
    import/no-extraneous-dependencies: ["error", {"devDependencies": true }]
  */
  const { argv } = require('yargs');

  const watchLocalModules = require('./utils/watchLocalModules').default;
  const devHolocronCDN = require('./devHolocronCDN').default;
  const oneAppDevProxy = require('@americanexpress/one-app-dev-proxy');
  const oneAppDevCdnPort = process.env.HTTP_ONE_APP_DEV_CDN_PORT;
  const oneAppDevProxyPort = process.env.HTTP_ONE_APP_DEV_PROXY_SERVER_PORT;

  serverChain = Promise.resolve()
    .then(() => new Promise((res, rej) => addServer(
      devHolocronCDN.listen(oneAppDevCdnPort, (err) => {
        if (err) {
          rej(err);
        } else {
          console.log(`ðŸ‘• one-app-dev-cdn server listening on port ${oneAppDevCdnPort}`);
          res();
        }
      })
    )))
    .then(() => new Promise((res, rej) => addServer(
      oneAppDevProxy({
        useMiddleware: argv.m,
        pathToMiddleware: path.join(__dirname, '../../.dev/middleware'),
        remotes: getRemotesFromDevEndpointsFile(),
      }).listen(oneAppDevProxyPort, (err) => {
        if (err) {
          rej(err);
        } else {
          console.log(`ðŸ‘– one-app-dev-proxy server listening on port ${oneAppDevProxyPort}`);
          res();
        }
      })
    )))
    .then(appServersStart)
    .then(watchLocalModules);
} else {
  serverChain = appServersStart();
}

export default serverChain.catch((err) => {
  console.error(err);
  shutdown();
});
