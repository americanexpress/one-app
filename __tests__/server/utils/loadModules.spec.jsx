/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { getModule } from 'holocron';
import { updateModuleRegistry } from 'holocron/server';
import { CONFIGURATION_KEY } from '../../../src/server/utils/onModuleLoad';
import loadModules from '../../../src/server/utils/loadModules';
import { updateCSP } from '../../../src/server/middleware/csp';
import { setClientModuleMapCache, getClientModuleMapCache } from '../../../src/server/utils/clientModuleMapCache';

// This named export exists only on the mock
// eslint-disable-next-line import/named

jest.mock('holocron');
jest.mock(
  'holocron/server',
  () => ({
    updateModuleRegistry: jest.fn(),
  })
);
jest.mock('../../../src/server/utils/stateConfig', () => ({
  getServerStateConfig: jest.fn(() => ({
    rootModuleName: 'some-root',
  })),
}));

jest.mock('../../../src/server/middleware/csp', () => ({
  updateCSP: jest.fn(),
}));

const RootModule = () => ({});

describe('loadModules', () => {
  const moduleMap = {
    modules: {
      'some-root': {
        node: {
          url: 'https://example.com/cdn/some-root/2.2.2/some-root.node.js',
          integrity: '4y45hr',
        },
        browser: {
          url: 'https://example.com/cdn/some-root/2.2.2/some-root.browser.js',
          integrity: 'nggdfhr34',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/some-root/2.2.2/some-root.legacy.browser.js',
          integrity: '7567ee',
        },
      },
    },
  };

  beforeAll(() => {
    getModule.mockImplementation(() => RootModule);
    updateModuleRegistry.mockImplementation(() => ({
      'some-root': {
        node: {
          url: 'https://example.com/cdn/some-root/2.2.2/some-root.node.js',
          integrity: '4y45hr',
        },
        browser: {
          url: 'https://example.com/cdn/some-root/2.2.2/some-root.browser.js',
          integrity: 'nggdfhr34',
        },
        legacyBrowser: {
          url: 'https://example.com/cdn/some-root/2.2.2/some-root.legacy.browser.js',
          integrity: '7567ee',
        },
      },
    }));

    global.fetch = jest.fn(() => Promise.resolve({
      json: () => Promise.resolve(moduleMap),
    }));
  });

  beforeEach(() => {
    jest.clearAllMocks();
    setClientModuleMapCache({ modules: {} });
  });

  it('updates the holocron module registry', async () => {
    await loadModules();
    expect(updateModuleRegistry).toHaveBeenCalledWith({
      moduleMap,
      batchModulesToUpdate: require('../../../src/server/utils/batchModulesToUpdate').default,
      getModulesToUpdate: require('../../../src/server/utils/getModulesToUpdate').default,
      onModuleLoad: require('../../../src/server/utils/onModuleLoad').default,
    });
  });

  it('updates the client module map cache', async () => {
    await loadModules();
    expect(getClientModuleMapCache()).toMatchSnapshot();
  });

  it('doesnt update caches when there are no changes', async () => {
    updateModuleRegistry.mockImplementationOnce(() => ({}));
    await loadModules();
    expect(getClientModuleMapCache()).toMatchSnapshot();
  });

  it('updates CSP when csp is set on root module if root module is loaded', async () => {
    RootModule[CONFIGURATION_KEY] = {
      csp: "default-src 'none';",
    };

    await loadModules();
    expect(updateCSP).toHaveBeenCalledWith("default-src 'none';");
  });

  it('calls updateCSP even when csp is not set', async () => {
    delete RootModule[CONFIGURATION_KEY].csp;
    await loadModules();
    expect(updateCSP).toHaveBeenCalledWith(undefined);
  });

  describe('when root module not loaded', () => {
    beforeAll(() => {
      updateModuleRegistry.mockImplementation(() => ({
        'not-a-root': {
          node: {
            url: 'https://example.com/cdn/not-a-root/2.2.2/not-a-root.node.js',
            integrity: '4y45hr',
          },
          browser: {
            url: 'https://example.com/cdn/not-a-root/2.2.2/not-a-root.browser.js',
            integrity: 'nggdfhr34',
          },
          legacyBrowser: {
            url: 'https://example.com/cdn/not-a-root/2.2.2/not-a-root.legacy.browser.js',
            integrity: '7567ee',
          },
        },
      }));
    });

    it('does not attempt to update CSP', async () => {
      await loadModules();
      expect(updateCSP).not.toHaveBeenCalled();
    });
  });
});
