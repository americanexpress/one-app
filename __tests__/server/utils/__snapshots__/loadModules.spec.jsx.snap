// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadModules doesnt update caches when there are no changes 1`] = `
Object {
  "browser": Object {
    "modules": Object {},
  },
  "legacyBrowser": Object {
    "modules": Object {},
  },
}
`;

exports[`loadModules updates the client module map cache 1`] = `
Object {
  "browser": Object {
    "modules": Object {
      "some-root": Object {
        "baseUrl": "https://example.com/cdn/some-root/2.2.2/",
        "browser": Object {
          "integrity": "nggdfhr34",
          "url": "https://example.com/cdn/some-root/2.2.2/some-root.browser.js",
        },
      },
    },
  },
  "legacyBrowser": Object {
    "modules": Object {
      "some-root": Object {
        "baseUrl": "https://example.com/cdn/some-root/2.2.2/",
        "legacyBrowser": Object {
          "integrity": "7567ee",
          "url": "https://example.com/cdn/some-root/2.2.2/some-root.legacy.browser.js",
        },
      },
    },
  },
}
`;
